[Docs] Add FAQ (#278)
exercism/v3: docs/maintainers/faq.md

# Frequently Asked Questions

## What is v3?

> v3's main goal is to improve the learning experience for students. We'll do this by explicitly designing exercises to teach Concepts. We call these Concept Exercises. Their goal is to teach one Concept, and that is all they should aim to teach. Concept Exercises are thus designed with a specific purpose in mind, whereas in v2 the learning experience was more incidental.

Reviewer: ```suggestion
v3's main goal is to improve the learning experience for students, so that when they complete a track they will be able to program in that language in an idiomatic way.

Exercism v2 currently focuses heavily on discovering lots of various information from a relatively complex exercise - something which can be challenging and confusing both for students and those trying to mentor them. In contrast, in v3, we are explicitly designing exercises to teach these Concepts. These new exercises are called Concept Exercises. Their goal is to teach one just Concept, providing enough challenge to familiarse the student with the idea of the Concept and enough mental scaffolding for them to know where to discover more information. Concept Exercises are therefore designed with a specific purpose in mind, whereas in v2 the learning experience was more incidental.

v3 also adds new features and changes to the user experience, such as In-browser coding, Approaches, better automated-feedback, and an improved mentoring system.
```

For a more in-depth explanation on v3, please check the following links:

- [Introduction to v3][readme]: An introduction to v3.
- [The features of v3][features]: A deeper dive into the changes we're making to v3.
- [Rationale for v3][rationale]. An explanation about the rationale behind some of our our design decisions for v3.

## What is the role of maintainers in the v3 transition?

> The key role for maintainers is in maintaining their tracks. We don't expect maintainers to do all the work of creating everything. We want to crowd-source work as much as possible. Maintainers are there to help guide, design and review, and do some of the actual exercise creation (but not all).

Reviewer: ```suggestion
The key role for maintainers is in designing and managing their tracks. We don't expect maintainers to do all the work of creating everything. We want to crowd-source work as much as possible. Maintainers' key role is in working out how their track can best teach a student, and designing and reviewing exercises exercises. We hope that maintainers will also enjoy creating some of the actual exercise themselves, but the load of this should be spread amongst the wider community.
```

## How many Concept Exercises should be created?

We're aiming for about 50 Concept Exercises, although that does depend on the language and how many concepts it has (some languages are more complex than others). There is no fixed number here, it all depends on the language and how many Concepts it has. For the more complex languages with lots of concepts, the number could easily go to 100 Concept Exercises.

As each Concept Exercise will have detailed instructions in the form of a GitHub issue, we're hoping to crowd-source the actual implementation of the Concept Exercises. Designing the Concept Exercises and creating the corresponding GitHub issues is a responsibility of the maintainers though.

## Is there a 1-1 mapping from Concept to Concept Exercise?

In general: yes. However, for very complex Concepts, it can make sense to have multiple smaller Concept Exercises instead of one huge Concept Exercise.

We highly recommend reading [this thread][concept-to-exercise-mapping] for a nice discussion on this topic.

## How will mentoring work for Concept Exercises?

> There will be no mentoring on Concept Exercises in v3. The process will be fully automated, and will work as follows:

Reviewer 2: Should there be some placeholder text here, to show that "as follows:" doesn't have the necessary information yet, but we expect it to be written?  Perhaps "TO BE WRITTEN" or something so that it brings attention to itself.

Reviewer 2: Or is it only appearing to be that way because the heading level of the next section(s) is the same as this section?

Author: Good catch. The second heading should be removed. Will update.

> ## How will students work with Concept Exercises?

Reviewer 2: I am becoming more convinced that the heading level here is the contributing factor to the feeling of the last paragraph leaving us hanging.

```suggestion
### How will students work with Concept Exercises?
```

Author: Agreed. The heading has been removed.

> 1. The student submits a solution for a Concept Exercise which prequisites have been met by the student (i.e. the student has completed Concept Exercises teaching the prerequisite Concepts).

Reviewer 2: The wording is confusing here, perhaps from a missing word?

```suggestion
1. The student submits a solution for a Concept Exercise for which prequisites have been met by the student (i.e. the student has completed Concept Exercises teaching the prerequisite Concepts).
```

Author: I've updated slightly. I'm not sure the added "for" helps here, as the prerequisites are linked to the exercise.

> 1. The solution is tested automatically to verify if it passes all tests. This is done by the [test runner][test-runner]. If a test fails, details of the failing test are shown to the user. The student has to submit again (step 1) in order to continue.

Reviewer 2: ```suggestion
1. The solution is tested automatically to verify it passes all tests. This is done by the [test runner][test-runner]. If a test fails, details of the failing test are shown to the user. The student must submit again (step 1) in order to continue.
```

Author: Thanks!

> 1. The [representer][representer] creates a representation of the solution. If there are any comments linked to that representation, those comments are presented to the student. The student has to submit again (step 1) in order to continue.

Reviewer 2: ```suggestion
1. The [representer][representer] creates a representation of the solution. If there are any comments linked to that representation, those comments are presented to the student. The student must submit again (step 1) in order to continue.
```

> 1. The [analyzer][analyzer] analyzes the solution. If there are comments, those comments are presented to the student. If one of the comments indicates that the student should not be allowed to proceed, the student has to submit again (step 1) in order to continue.

Reviewer 2: ```suggestion
1. The [analyzer][analyzer] analyzes the solution. If there are comments, those comments are presented to the student. If one of the comments indicates that the student should not be allowed to proceed, the student must submit again (step 1) in order to continue.
```

## What is the role of maintainers in the v3 transition?

The key role for maintainers is in maintaining their tracks. We don't expect maintainers to do all the work of creating everything. We want to crowd-source work as much as possible. Maintainers are there to help guide, design and review, and do some of the actual exercise creation (but not all).

For more information, see the [maintainers README][maintainers].

## What are Concept Exercises?

A Concept Exercise is an exercise that is designed to teach one Concept, and nothing more. Concept Exercises are thus designed with a specific purpose in mind, whereas in v2 the learning experience was more incidental.

As opposed to the v2 exercises, Concept Exercises are _not_ based on shared data anymore. The reason for that is that we don't want any cross-track dependency issues, where changing the shared data makes sense for some tracks, but not for other tracks. The Concept Exercises can thus be tailored completely to best fit their track.

For more information, see the [concept exercises document][concept-exercises].

## What is a Concept?

> A Concept describes a feature of a programming language. Examples are "Encapsulation", "Conditionals", and "Functions." In other words, Concepts are what one would need to learn if one was to learn a new (programming) language.

Reviewer: ```suggestion
A Concept describes a feature of a programming language. Examples are "Encapsulation", "Conditionals", and "Functions." In other words, Concepts are what one would need to learn if one was to learn a new (programming) language.

Although there will obviously be cross-over between languages, Concepts are indented to be language-specific. What is it about **your** language that a junior programmer would need to learn, unlearn and remap to be able to successfully transition to it.
```
